src/main/scala/com/fulcrumgenomics/umi/Umis.scala

```diff
@@ -35,37 +35,62 @@ object Umis {
     * The read name is split by the given name delimiter, and the last field is assumed to be the UMI sequence.  The UMI
     * will be copied to the `RX` tag as per the SAM specification.
     *
+    * An exception will be thrown if the read name does not contain a valid UMI in the last delimited segment.
+    *
     * @param rec the record to modify
     * @param removeUmi true to remove the UMI from the read name, otherwise only copy the UMI to the tag
-    * @param umiDelimiter if not None, replaces any occurrences of this delimiter found in the UMI with a dash ('-')
-    *                     as per the SAM specification
+    * @param delimiter the delimiter of fields within the readname
     * @return the modified record
     */
-  def copyUmiFromReadName(rec: SamRecord,
-                          removeUmi: Boolean = false,
-                          nameDelimiter: Char = ':',
-                          umiDelimiter: Option[Char] = Some('+')): SamRecord = {
-    // extract the UMI
-    val idx = rec.name.lastIndexOf(nameDelimiter)
-    require(idx != -1, s"Read did not have multiple '$nameDelimiter'-separated fields: ${rec.name}")
-    val rawUmi = rec.name.substring(idx + 1, rec.name.length)
+  def copyUmiFromReadName(rec: SamRecord, removeUmi: Boolean = false, delimiter: Char = ':'): SamRecord = {
+    // Extract and set the UMI
+    val umi = extractUmisFromReadName(rec.name, delimiter, strict=false)
+    require(umi.nonEmpty, f"No valid UMI found in: ${rec.name}")
+    umi.foreach(u => rec(ConsensusTags.UmiBases) = u)
 
-    // re-delimit the UMI if desired, validating as necessary
-    val umi = umiDelimiter match {
-      case Some(delim) => rawUmi.replace(delim, '-')
-      case None        => rawUmi
-    }
+    // Remove the UMI from the read name if requested
+    if (removeUmi) rec.name = rec.name.substring(0, rec.name.lastIndexOf(delimiter))
 
-    // validate the UMI
-    require(umi.nonEmpty && umi.forall { char => SequenceUtil.isValidBase(char.toByte) || SequenceUtil.isNoCall(char.toByte) || char == '-' },
-      f"UMI '$umi' is not valid (not `[ACGTNacgtn-]+`) for read: ${rec.name}"
-    )
+    rec
+  }
+
+  /**
+    * Extracts the UMI from an Illumina fastq style read name.  Illumina documents their FASTQ read names as:
+    *   @<instrument>:<run number>:<flowcell ID>:<lane>:<tile>:<x-pos>:<y-pos>:<UMI> <read>:<is filtered>:<control number>:<index>
```

Comment:
UMI is optional no?  

Also document that dual-UMIs should be plus delimited

Reply:
That's a direct copy/paste from the Illumina doc.  I'll add context but prefer not to edit the actual line.